### DISCUSSION TITLE
`set` type is too lax and allow setting invalid state

### DISCUSSION THREAD

**Discussion by pawbor** [pmndrs/zustand#1864] [General]
## Summary
Hi, first of all big thanks for all the hard work put into creating this library!

TBH I'm not sure if this should be considered a bug. I guess it's a design decision, but because of it, it is possible to set state that does not match types defined with typescript.

The cause of the problem is the ability to pass partial state to `set`. Lets consider types below:
```
type A = {
  type: "a";
};

type B = {
  type: "b";
  prop: string;
};

type State = A | B;
```
When transitioning from `A` to `B` it's possible to do `set({type: 'b'})`. This will lead to invalid state since we did not define `prop`. Unfortunately typescript will not catch that since it tests for `Partial<A | B>`.

## Link to reproduction
https://codesandbox.io/s/charming-nobel-6gqez5?file=/src/App.tsx

## Check List

Please do not ask questions in issues.

- [ ] I've already opened a [discussion](https://github.com/pmndrs/zustand/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [x] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

Please check this if you're filing an issue regarding TypeScript.

- [x] I've read the [typescript guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md), in particular that `create` is to be used as `create<T>()(...)` and not `create<T>(...)`.

**Comment by dai-shi** [comment]
Does enabling `exactOptionalPropertyTypes` change the behavior?
https://github.com/pmndrs/zustand/blob/820dbd35eabc36d0117e3ac9ee9c1cb36336e88e/docs/migrations/migrating-to-v4.md?plain=1#L123-L132

**Comment by pawbor** [comment]
Unfortunately not.
`exactOptionalPropertyTypes` forbids you to explicitly assign `undefined` to a optional property if it is defined, but you can still skip defining it. 
So `set({type: 'b'})` does not throw because it does not define the `prop` at all, and it is optional because of `Partial`. 
But `set({type: 'b', prop: undefined })` throws an error because if `prop` is defined then it must be a `string`.

https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true#code/C4TwDgpgBAQlC8UDeAoKVSQFxQEQCNcBuNKMAJwHswcBnYcgSwDsBzEgXxJQHoeoAogA8wAG0YBjRsCgBXZgBMIAMxYQFGABZUA7rSgRyVcigmVm9KPgCMOAAoBDcsEYPRAHhgA+BMlKYIHAJcABpSCmoceSVVZnUULhRefgADCLAUqGZKGRi1DVpKKGkAcn1KAGtTc0t8ACZ7Jxc3Tx9EVHQAoMIwxOSoNKoM4v16JjYoQuLgMqhK6osZfABmRudXD29fDoxwQLwe8KGcEpKE7j5BEXEpGWiVfK1dfUNjFBsAOnTfe9j1C9S6UyjFGDBYrEmRVK5Sqn2+iFOJCAA

**Comment by dai-shi** [comment]
Ah, I misunderstood it.
This can also break typing: https://tsplay.dev/mxkY8N

I'm not sure if it's possible to fix.
If anyone is interested, please give it a try.
There were some discussions in the past to decide we prefer relying on `exactOptionalPropertyTypes`, which might be related or might not in this case.


I'll leave this issue open for a while.

**Comment by JacobWeisenburger** [comment]
Sounds like a job for [Zod](https://github.com/colinhacks/zod)

**Comment by pawbor** [comment]
tbh not sure how using zod would be a solution here. the whole point of defining types in zustand store is to have a "compilation" time safety instead of validating on runtime. As I mentioned, the flaw in zustand types is that it allows setting partial state and that may lead to invalid state as I described above.

The "simple" solution would be to enforce providing full state instead of just part of it, similar to how it's done in redux reducers. Unfortunately this would be a serious breaking change and I guess that this was a design decision to make nicer developer experience.

**Comment by ScottAwesome** [comment]
> tbh not sure how using zod would be a solution here. the whole point of defining types in zustand store is to have a "compilation" time safety instead of validating on runtime. As I mentioned, the flaw in zustand types is that it allows setting partial state and that may lead to invalid state as I described above.

You could do zod validation in development and testing environments (staging etc) and strip them for production builds using build flags if your bundler or toolchain supports interpolating a global (webpack, vite, terser, rollup, esbuild, swc and many others have support for this)

Maybe not ideal, but it'd strip weight from prod and allow you to enforce with more rigor otherwise

**Comment by bhirsch42** [comment]
I wrote a wrapper that creates a zustand store with better type guarantees:

```ts
import { StoreApi, createStore } from "zustand/vanilla";

type SafeSet<T> = (state: T | ((state: T) => T)) => void;

type SafeStoreApi<T> = Omit<StoreApi<T>, "setState"> & {
  setState: SafeSet<T>;
};

function createSafeStore<T>(
  initializer: (set: SafeSet<T>, get: () => T, store: SafeStoreApi<T>) => T
): SafeStoreApi<T> {
  return createStore(initializer);
}

type DiscriminatedUnionStore =
  | {
      type: "data";
      value: number;
    }
  | { type: "error"; message: string };

const store = createSafeStore<DiscriminatedUnionStore>((set) => ({
  type: "data",
  value: 0,
}));

store.setState(() => ({ type: "error" })); // TypeError
```

This approach feels cumbersome to use with middleware, since the middleware calls have to go inside of `createSafeStore`. Could be verbose if you have different middleware for different stores. 

---

Another approach could be to initialize every store with a `value` prop, so all your state is nested, so it's not affected by the `Partial` typing.

Ex. Instead of this:
```ts
type Counter = {
  count: number;
  increment(by: number): void;
};

const useStore = create<Counter>((set) => ({
  count: 0,
  increment(by) {
    set((state) => ({ count: state.count + by }));
  },
}));
```

Do this:
```ts
type Counter = {
  value: { count: number };
  increment(by: number): void;
};

const useStore = create<Counter>((set) => ({
  value: { count: 0 },
  increment(by) {
    set((state) => ({ value: { count: state.value.count + by } }));
  },
}));
```
